src/test/crimson/test_messenger.cc

```diff
@@ -54,19 +54,18 @@ static entity_addr_t get_server_addr() {
   return saddr;
 }
 
+template <typename T>
+static std::list<seastar::lw_shared_ptr<seastar::sharded<T>>> sharded_objects;
+
 template <typename T, typename... Args>
 seastar::future<T*> create_sharded(Args... args) {
-  // we should only construct/stop shards on #0
-  return seastar::smp::submit_to(0, [=] {
+  return seastar::smp::submit_to(0, [=]() -> seastar::future<seastar::sharded<T>*> {
     auto sharded_obj = seastar::make_lw_shared<seastar::sharded<T>>();
-    return sharded_obj->start(args...
-    ).then([sharded_obj] {
-      seastar::engine().at_exit([sharded_obj] {
-        return sharded_obj->stop().then([sharded_obj] {});
-      });
+    sharded_objects<T>.push_back(sharded_obj);
+    return sharded_obj->start(args...).then([sharded_obj] {
```

Comment:
We should also probably stop from shard 0, right?
deferred_stop can be used in a seastar::async thread which would be submitted to as well in this scope.
Alternatively, instead of `seastar::parallel_for_each` we should probably `invoke_on(0 ...` to keep the same logic.